File: wpimath/src/test/java/org/wpilib/math/trajectory/proto/DifferentialTrajectoryProtoTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.trajectory.proto;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import org.junit.jupiter.api.Test;
import org.wpilib.math.geometry.Pose2d;
import org.wpilib.math.geometry.Rotation2d;
import org.wpilib.math.kinematics.ChassisAccelerations;
import org.wpilib.math.kinematics.ChassisVelocities;
import org.wpilib.math.trajectory.DifferentialSample;
import org.wpilib.math.trajectory.DifferentialTrajectory;

class DifferentialTrajectoryProtoTest {
  private static final DifferentialTrajectory DATA =
      new DifferentialTrajectory(
          List.of(
              new DifferentialSample(
                  0.0,
                  Pose2d.kZero,
                  new ChassisVelocities(1.0, 0.0, 0.0),
                  new ChassisAccelerations(0.5, 0.0, 0.0),
                  0.9,
                  1.1),
              new DifferentialSample(
                  1.0,
                  new Pose2d(1.1, 2.2, Rotation2d.fromDegrees(30)),
                  new ChassisVelocities(1.5, 0.0, 0.3),
                  new ChassisAccelerations(0.1, 0.0, 0.4),
                  1.4,
                  1.6)));

  @Test
  void testRoundtrip() {
    var proto = DifferentialTrajectory.proto.createMessage();
    DifferentialTrajectory.proto.pack(proto, DATA);

    DifferentialTrajectory data = DifferentialTrajectory.proto.unpack(proto);
    assertEquals(DATA.getSamples(), data.getSamples());
  }
}

File: wpimath/src/test/native/cpp/trajectory/proto/DifferentialTrajectoryProtoTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "wpi/math/trajectory/DifferentialTrajectory.hpp"
#include "wpi/util/SmallVector.hpp"

using namespace wpi::math;

namespace {
const DifferentialTrajectory kExpectedData =
    DifferentialTrajectory{std::vector<DifferentialSample>{
        DifferentialSample{
            0_s, Pose2d{}, ChassisVelocities{1_mps, 0_mps, 0_rad_per_s},
            ChassisAccelerations{0.5_mps_sq, 0_mps_sq, 0_rad_per_s_sq}, 0.9_mps,
            1.1_mps},
        DifferentialSample{
            1_s, Pose2d{1.1_m, 2.2_m, 30_deg},
            ChassisVelocities{1.5_mps, 0_mps, 0.3_rad_per_s},
            ChassisAccelerations{0.1_mps_sq, 0_mps_sq, 0.4_rad_per_s_sq},
            1.4_mps, 1.6_mps}}};
}  // namespace

TEST_CASE("DifferentialTrajectoryProtoTest Roundtrip", "[wpimath]") {
  wpi::util::ProtobufMessage<decltype(kExpectedData)> message;
  wpi::util::SmallVector<uint8_t, 64> buf;

  REQUIRE(message.Pack(buf, kExpectedData));
  auto unpacked_data = message.Unpack(buf);
  REQUIRE(unpacked_data.has_value());

  CHECK(kExpectedData == unpacked_data.value());
}
